src/gpio.rs

```diff
@@ -59,498 +63,197 @@
 //! [InputPin]: embedded_hal::digital::v2::InputPin
 //! [OutputPin]: embedded_hal::digital::v2::OutputPin
 //! [examples/toggle.rs]: https://github.com/stm32-rs/stm32f3xx-hal/blob/v0.9.0/examples/toggle.rs
+//!
+//! If you need a more temporary mode change, and can not use the `into_<mode>` functions for
+//! ownership reasons, you can use the closure based `with_<mode>` functions to temporarily change the pin type, do
+//! some output or input, and then have it change back once done.
+//!
+//! ### Dynamic Mode Change
+//! The above mode change methods guarantee that you can only call input functions when the pin is
+//! in input mode, and output when in output modes, but can lead to some issues. Therefore, there
+//! is also a mode where the state is kept track of at runtime, allowing you to change the mode
+//! often, and without problems with ownership, or references, at the cost of some performance and
+//! the risk of runtime errors.
+//!
+//! To make a pin dynamic, use the `into_dynamic` function, and then use the `make_<mode>` functions to
+//! change the mode
+
+#![allow(missing_docs)]
+
+use core::marker::PhantomData;
+
+use crate::pac::{Interrupt, EXTI};
+use crate::rcc::AHB;
+use crate::Toggle;
 
-use core::{convert::Infallible, marker::PhantomData};
+mod convert;
+use convert::PinMode;
+mod partially_erased;
+pub use partially_erased::{PEPin, PartiallyErasedPin};
+mod erased;
+pub use erased::{EPin, ErasedPin};
+mod dynamic;
+pub use dynamic::{Dynamic, DynamicPin};
+mod hal_02;
```

Comment:
Hm, `hal_02` seem a little cryptic for a module name. Maybe use the conventional abbreviation of `embeded_hal` like `eh`

Or as it is not typed very often I'm fine with a long name like `embedded_hal_v0_2` or similar.